### ISSUE TITLE
Result difference when running beam search on ONNX T5 model.

### ISSUE THREAD

**Issue by VikasOjha666** [microsoft/onnxruntime#10942]
**Describe the bug**
I converted a T5 base model to ONNX and I implemented the IO bindings method for inference. When I ran the greedy search using the hugging face library it was faster as my t5_decoder was in fp16 precision. But when I used beam search it was slower. Hence I decided to implement beam search from scratch. I also implemented greedy search from scratch as well and in that case, the model was working fine. But in the case of custom beam search implementation, I am observing a strange issue. The issue is that the result of the decoding was garbage. But when I just add a delay in the beam search loop everything works as expected.

**Urgency**
It's very urgent as a similar model is running in production and incurring huge costs which can be saved by optimization.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- ONNX Runtime installed from (source or binary): Tried with both from PyPI and by building from source.
- ONNX Runtime version: 1.11
- Python version: 3.7.12
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:11.1
- GPU model and memory: Tried with both K80 and T4

This is the code for beam search implementation. Although it's not a complete implementation.

#Encoder prediction.
dec_input_ids=torch.tensor([[0],[0],[0],[0]],device=device)
enc_out=t5_enc(input_ids.repeat_interleave(4, dim=0))
dec_out=t5_dec(dec_input_ids,enc_out)

b_batch_size=input_ids.shape[0]
batch_size=input_ids.shape[0]
beam_scores = torch.zeros((b_batch_size, num_beams), dtype=torch.float, device=device)
beam_scores[:, 1:] = -1e9
beam_scores = beam_scores.view((batch_size * num_beams,))

for _ in range(10):
  dec_out=t5_dec(dec_input_ids,enc_out)
  next_token_logits = dec_out[:, -1, :]
  next_token_scores = torch.nn.functional.log_softmax(
                    next_token_logits, dim=-1
                )  # (batch_size * num_beams, vocab_s)
  ####time.sleep(0.00825607776)
  next_token_scores_processed = logits_processor(input_ids, next_token_scores)
  next_token_scores = next_token_scores_processed + beam_scores[:, None].expand_as(next_token_scores)
  vocab_size = next_token_scores.shape[-1]
  next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
  next_token_scores, next_tokens = torch.topk(
                    next_token_scores, num_beams, dim=1, largest=True, sorted=True
                )
  next_indices = torch_int_div(next_tokens, vocab_size)
  next_tokens = next_tokens % vocab_size
 
  next_beam_scores,next_beam_tokens,next_beam_indices=beam_search_scorer(next_tokens,next_token_scores,next_indices)
  #####time.sleep(0.00825607776)

  beam_scores = next_beam_scores
    
  beam_next_tokens = next_beam_tokens
  beam_idx = next_beam_indices

  dec_input_ids = torch.cat([dec_input_ids[beam_idx, :], beam_next_tokens.unsqueeze(-1)], dim=-1)

With time.sleep() commented the output is:
['What) gravity ) which',
 'Whatweight ores',
 'Whatweight or is',
 'Whatweight orss']

With time.sleep() executing output is corrected as expected.
['What is a natural phenomenon by which all things',
 'What is the natural phenomenon by which all things with',
 'What is the name of the natural phenomenon by which',
 'What is a natural phenomenon by which all objects']

@ManuelAngel99 @mtodd @mattetti @sverrejoh @bgianfo

**Comment by tianleiwu** [comment]
If adding sleep could solve the issue. It is possible that there is data race, and replace sleep by cuda synchronize might fix the issue.

**Comment by VikasOjha666** [comment]
@tianleiwu Thanks for your time.Yes I also suspect there is data race issues. In IO bindings I tried different ways. In one way I built onnxruntime from source and just did the IO binding to input and then returned the ORT value from session later in the loop part I converted the ORT to torch tensor. In other way I pre created a torch tensor with output size and binded it's memory to ORT output. Both of these had the same issue. Disabling bindings tends to solve the issue but it slower than the binding due to data copy between CPU and GPU. Will synchronise CUDA might solve it?

**Comment by VikasOjha666** [comment]
@tianleiwu Can you suggest any reference for implementing cuda synchronise?

**Comment by VikasOjha666** [comment]
@tianleiwu Did you mean torch.cuda.synchronise() method?

**Comment by VikasOjha666** [comment]
The problem was data race. Adding torch.cuda.synchronise(device="cuda") solved the issue.
